mavros_extras/src/plugins/odom.cpp

```diff
@@ -31,103 +31,42 @@ using Matrix6d = Eigen::Matrix<double, 6, 6, Eigen::RowMajor>;
 /**
  * @brief Odometry plugin
  *
- * Sends odometry data to the FCU position and attitude estimators.
- * @see odom_cb()
+ * Sends odometry data to the FCU position and attitude estimators and
+ * publishes odometry data that comes from FCU.
+ *
+ * This plugin is following ROS REP 147. Pose is expressed in parent frame. (Quaternion rotates from child to parent)
+ * The twist is expressed in the child frame.
+ *
+ * @see odom_cb()	transforming and sending odometry to fcu	-> IN  direction
+ * @see handle_odom()	receiving and transforming odometry from fcu	-> OUT direction
  */
 class OdometryPlugin : public plugin::PluginBase {
 public:
-	EIGEN_MAKE_ALIGNED_OPERATOR_NEW 	// XXX(vooon): added to try to fix #1223. Not sure that it is needed because class do not have Eigen:: fields.
+	EIGEN_MAKE_ALIGNED_OPERATOR_NEW		// XXX(vooon): added to try to fix #1223. Not sure that it is needed because class do not have Eigen:: fields.
 
 	OdometryPlugin() : PluginBase(),
 		odom_nh("~odometry"),
-		body_frame_orientation_in_desired("flu"),
-		body_frame_orientation_out_desired("frd"),
-		child_frame_id("base_link"),
-		frame_id("odom"),
-		local_frame_in("local_origin_ned"),
-		local_frame_out("vision_ned")
+		fcu_odom_parent_id_des("map"),
```

Comment:
Why the `_des`?

Reply:
We receive odometry message with a parent and a child frame. These child frames are most likely not the ones that we want to send to fcu or to ROS. So will rotate the odometry message components such that they are in the right frame. Notice here that we only apply the rotation and do not take into account any translation. This is handled in the ECL/EKF

Reply:
> These child frames are most likely not the ones that we want to send to fcu or to ROS.

Reply:
> These child frames are most likely not the ones that we want to send to fcu or to ROS.

Why not? For example, my VIO system publishes an odometry message with parent_frame_id = mission, child_frame_id = base_link. I want to send this to the FCU. What is the use-case for not wanting to send this?

>  Notice here that we only apply the rotation and do not take into account any translation. This is handled in the ECL/EKF

Also this sounds wrong. If you're going to use the TF lookup, you should be applying the full transform, not just the rotation. Can you explain the rationale here?

Reply:
> Also this sounds wrong. If you're going to use the TF lookup, you should be applying the full transform, not just the rotation. Can you explain the rationale here?

The linear offset of the sensor is compensated in the EKF2 by the EKF_EV_POS_* params. Don't think about these rotation as transforming the odometry message, but rather as changing the frame orientation that is set by the creator of the VIO system. E.g. you could design your VIO system to have a world z axis that is not aligned with gravity, why ever one want to do so.

Reply:
> E.g. you could design your VIO system to have a world z axis that is not aligned with gravity, why ever one want to do so.

This is not something the plugin should be dealing with, in my opinion. This should be the bare minimum needed to translate between ROS and the mavlink message. If you system has such a weird convention of not using a gravity-aligned reference frame, I would expect you to pre-rotate the measurement before sending it into the plugin.